domain/src/main/java/org/oppia/domain/topic/StoryProgressController.kt

```diff
@@ -103,178 +81,117 @@ class StoryProgressController @Inject constructor(
-    return dataProviders.convertToLiveData(getTopicProgressDataProvider(profileId, topicId))
-  }
-
-  /** Returns story progress on per-profile basis specified by topicId and storyId. */
-  fun getStoryProgress(
-    profileId: ProfileId,
-    topicId: String,
-    storyId: String
-  ): LiveData<AsyncResult<StoryProgress>> {
-    return dataProviders.convertToLiveData(getStoryProgressDataProvider(profileId, topicId, storyId))
-  }
-
-  /** Returns chapter progress on per-profile basis specified by topicId, storyId and explorationId. */
-  fun getChapterProgress(
-    profileId: ProfileId,
-    topicId: String,
-    storyId: String,
-    exploration: String
-  ): LiveData<AsyncResult<ChapterPlayState>> {
-    return dataProviders.convertToLiveData(
-      getChapterProgressDataProvider(
-        profileId,
-        topicId,
-        storyId,
-        exploration
-      )
-    )
-  }
-
-  private fun getProfileProgressDataProvider(profileId: ProfileId): DataProvider<TopicProgressDatabase> {
+  /** Returns [TopicProgressList] [DataProvider] for a particular profile. */
+  internal fun retrieveTopicProgressListDataProvider(profileId: ProfileId): DataProvider<TopicProgressList> {
     return dataProviders.transformAsync(
-      TRANSFORMED_GET_PROFILE_PROGRESS_PROVIDER_ID,
+      TRANSFORMED_GET_STORY_PROGRESS_LIST_PROVIDER_ID,
       retrieveCacheStore(profileId)
-    ) {
-      if (it.topicProgressMap.keys.isNotEmpty()) {
-        AsyncResult.success(it)
+    ) { topicProgressDatabase ->
+      val topicProgressListBuilder = TopicProgressList.newBuilder()
+      topicProgressListBuilder.addAllTopicProgress(topicProgressDatabase.topicProgressMap.values)
+      if (topicProgressListBuilder.topicProgressList.isNotEmpty()) {
+        AsyncResult.success(topicProgressListBuilder.build())
       } else {
-        AsyncResult.failed(ProfileProgressNotFoundException("ProfileId: $profileId does not contain any progress"))
+        AsyncResult.failed(TopicProgressListNotFoundException("ProfileId: $profileId does not contain any topic progress"))
       }
     }
   }
 
-  private fun getTopicProgressDataProvider(profileId: ProfileId, topicId: String): DataProvider<TopicProgress> {
+  /** Returns [StoryProgressList] [DataProvider] for a particular profile. */
+  internal fun retrieveStoryProgressListDataProvider(profileId: ProfileId): DataProvider<StoryProgressList> {
     return dataProviders.transformAsync(
-      TRANSFORMED_GET_TOPIC_PROGRESS_PROVIDER_ID,
+      TRANSFORMED_GET_STORY_PROGRESS_LIST_PROVIDER_ID,
       retrieveCacheStore(profileId)
-    ) {
-      val topicProgress = it.topicProgressMap[topicId]
-      if (topicProgress != null) {
-        AsyncResult.success(topicProgress)
+    ) { topicProgressDatabase ->
+      val storyProgressListBuilder = StoryProgressList.newBuilder()
+      topicProgressDatabase.topicProgressMap.values.forEach { topicProgress ->
```

Comment:
I suggest transforming this into a new list and adding it to the StoryProgressList builder, instead (in general, I recommend against passing mutable state into a transformation lambda since that introduces a sync point which makes parallelizing the functional chain much harder). It's a bit cleaner to compute the chain in an immutable way and compile the results together at the end.

Reply:
@BenHenning I am not able to understand this.

Reply:
This is in relation to my earlier code examples. Rather than code like:

```kotlin
var builder = ...
collection.forEach(val -> {
  builder.add(transform(val))
})
return builder.build()
```

Prefer:

```kotlin
val transformedList = collection.map(val -> {
  transform(val)
})
return newBuilder().addAll(transformedList).build()
```

For the reasons mentioned above.

Reply:
Updated implementation to:
```
      val storyProgressListBuilder = StoryProgressList.newBuilder()
      val transformedStoryProgressList =
        topicProgressDatabase.topicProgressMap.values.map { topicProgress -> topicProgress!!.storyProgressMap!!.values }
      transformedStoryProgressList.map { storyProgressList ->
        storyProgressListBuilder.addAllStoryProgress(
          storyProgressList
        )
      }
```

Reply:
Ah, what I meant was not performing mutations inside of map functions. In order to achieve the benefits of functional programming, transform functions should have no side effects. That’s why in my example the builder bit happened at the end of the function, after the functional pieces are done. Another way to think about this: don’t change values in map functions, only return something new and use the final mapped result.

Reply:
Actually I got that but the `transformedList ` that I will be getting by that code will be `List<List<StoryProgress>>`, now finally we want list of strong progress so will need to iterate this one more time using some iterator.

Reply:
Okay got, it. I will use flatten for this.

Reply:
```
 val storyProgressListBuilder = StoryProgressList.newBuilder()
      val transformedStoryProgressList =
        topicProgressDatabase.topicProgressMap.values.map { topicProgress -> topicProgress!!.storyProgressMap!!.values.toList() }
      storyProgressListBuilder.addAllStoryProgress(transformedStoryProgressList.flatten())
```

This is correct right?

Reply:
Yup that’s what I had in mind.